src/lib.rs

```diff
@@ -388,23 +390,61 @@ where
 {
     info!("list-objects: bucket={}, key={}", bucket, key);
 
-    let mut bucket_contents = Vec::new();
-    let mut continuation: Option<String> = None;
-    loop {
-        let listing = list_objects(s3, bucket, key, continuation).await?;
-        if !listing.objects.is_empty() {
-            for entry in listing.objects {
+    let batches: Vec<_> = s3_stream_objects(s3, bucket, key).try_collect().await?;
+
+    let keys: Vec<_> = batches
+        .into_iter()
+        .flat_map(|batch| {
+            batch.into_iter().map(|entry| {
                 info!("key={}, etag={}", entry.key, entry.e_tag);
-                bucket_contents.push(entry.key);
+                entry.key
+            })
+        })
+        .collect();
+
+    Ok(keys)
+}
+
+pub fn s3_stream_objects<'a, T>(
```

Comment:
Should this be something like `s3_list_objects_stream`?  Which seems a bit verbose... so maybe we can combine with renaming the non-streaming version too... so

`s3_list_objects` -> `s3_list` (add a backward compat wrapper for the old name)
`s3_stream_objects` -> `s3_list_streaming`

If you know of a better convention for this, feel free to use that.

Reply:
Yeah I think that makes sense. Another way to shorten it could be to also remove `s3_` from all the functions, it does seem a little redundant. Although that would be a breaking change so maybe not worth it atm

Reply:
@notoriaga I'm fine with a breaking change if we add backward compat wrappers, or just rev the major version

Reply:
I removed all the `s3_` prefixes and I think it looks good. There was a name conflict then with some non-pub functions so those got renamed

```
head_object -> head_object_request
list_objects -> list_objects_request
```

`_request` meaning those are the functions at actually call the s3 api. `s3_stream_objects` is now `list_objects_stream`, which is a little wordy but also descriptive

Reply:
Also one small implementation change, the sync down function now uses `list_objects_stream` so we don't have to duplicated the logic for paginating the list request